File: Crio/Recursion/11_PopulatingNextRightPointersInEachNode/Solution.py
from crio.ds.Tree.TreeNode import TreeNode

# CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
def getLevelOrder(root, level, levelOrder):
    if not root:
        return
    levelOrder[level].append(root)
    getLevelOrder(root.left, level+1, levelOrder)
    getLevelOrder(root.right, level+1, levelOrder)
# CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

'''
# Definition of TreeNode
class TreeNode:
	def __init__(self, val):
		self.val = val
		self.left = None
		self.right = None
        self.next = None
'''
def populatingNextRightPointersInEachNode(root):
    # CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    levelOrder = [[] for i in range(20)]

    getLevelOrder(root, 0, levelOrder)
    for level in levelOrder:
        for i in range(len(level)-1):
            level[i].next = level[i+1]

    return root
    # CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

File: Crio/Recursion/11_PopulatingNextRightPointersInEachNode/Solution.java
import java.util.*;
import crio.ds.Tree.TreeNode;

/*
// Definition of TreeNode
public class TreeNode {
    public long val;
    public TreeNode left;
    public TreeNode right;
    public TreeNode next;

    public TreeNode (long x) {
        val = x;
        left = null;
        right = null;
        next = null;
    }
}
*/

public class Solution {
    public TreeNode populatingNextRightPointersInEachNode(TreeNode root) {
        // CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
        Queue<TreeNode> q = new  LinkedList<TreeNode>();
        if (root == null)
            return null;
        q.add(root);
        while (!q.isEmpty()) {
            int size = q.size();
            for (int i = 0 ; i < size ; i++) {
                TreeNode temp = q.peek();
                q.remove();
                if (temp.left != null) {
                    q.add(temp.left);
                }
                if (temp.right != null) {
                    q.add(temp.right);
                }
                if (i == size - 1) {
                    temp.next = null;
                } else {
                    temp.next = q.peek();
                }
            }
        }
        return root;
        // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
    }
}
